### ISSUE TITLE
Page visibility is cached beyond login/logout

### ISSUE THREAD

**Issue by ThrawnCA** [ckan/ckan#6955]
**CKAN version**

2.9

**Describe the bug**

Accessing a restricted page such as a private dataset will cache the page; returning to that page will initially show the cached version even if access has changed due to login/logout.

**Steps to reproduce**

* Log in as a privileged user.
* Go to a private dataset.
* Log out.
* Use the 'Back' button to return to the dataset. The page will initially be visible, but will deny access when reloaded.

This also works in reverse:

* Enter the URL of a private dataset while logged out. You will receive an error.
* Log in.
* Use the 'Back' button to return to the page. You will see the error again, until the page is reloaded.

**Expected behavior**

Access controls on a page should apply immediately after logging in/out.

**Additional details**

This becomes even more apparent with plugins like ckanext-datarequest; browsing to the data request list page, clicking the login button there, and logging in, will result in returning to a data request list page that looks logged out, still having a login button, until it is reloaded.

**Comment by Zharktas** [comment]
As far as I know, browsers don't reload the page when back button is pressed, they only execute the javascript that was already there. As CKAN uses server side rendered templates, instead of modern frontend rendering, there isn't much that can be done for cached pages.

**Comment by ThrawnCA** [comment]
The ckanext-datarequests example doesn't use the Back button, though...

**Comment by wardi** [comment]
@ThrawnCA If this is a concern for your site one fix could be to use response headers to disable caching for all pages when a user is logged in. Maybe that's what web sites do when they have a "I am logging in from a shared computer" option on their login screens.

**Comment by Zharktas** [comment]
The login workflow and cookie settings might also relate. If the workflow is doing proper redirects and you are using external authentication service, you need to use Lax in `who.samesite` as Strict prevents sending the cookie after redirects in external service.

**Comment by ThrawnCA** [comment]
There's no external authentication provider at https://www.data.qld.gov.au which is where we're encountering this.

**Comment by ThrawnCA** [comment]
Further investigation has revealed that CKAN uses "Cache-Control: private" whenever a user is logged in, meaning that pages will be cached in the browser but not publicly.

This ensures that sensitive pages will remain secure, but it does not address the issue of highly dynamic pages that must be loaded fresh every time, such as `/user/_logout`.

**Comment by ThrawnCA** [comment]
> @ThrawnCA If this is a concern for your site one fix could be to use response headers to disable caching for all pages when a user is logged in.

By the time someone is logged in, it may be too late; the browser may have cached an _un_authenticated version of a page, and will display it to them again, without ever getting a chance to send the no-cache header(s).

We have client complaints to that specific effect, actually, that they sometimes encounter a dataset page where they appear to be logged out. Then clicking on Login brings them to a login page that tells them they're already logged in. But using the Back button doesn't necessarily clear the cache and still says they're logged out on the dataset page...

**Comment by ThrawnCA** [comment]
@pdelboca I have devised a workaround to detect the major stale caching problem. When rendering the login link, it includes JavaScript that sends a HEAD request to the dashboard, to determine whether it's actually logged in already. If so, it alerts the user and offers to reload the page.

This should catch cases where someone is logged in but looking at a stale cache that says they're not logged in.

**Comment by pdelboca** [comment]
Thanks @ThrawnCA ! I agree with your comment that the whole `caching` system needs to be revisit and simplified.

I will bring this up to discussion in our next Development meeting and come back with some feedback.

Thanks!

**Comment by EricSoroos** [comment]
Could we add a "Varies: cookie" header so that the session cookie is part of the cache key?

**Comment by ThrawnCA** [comment]
> Could we add a "Varies: cookie" header so that the session cookie is part of the cache key?

I don't think so. Cookies are only sent on requests, IIUC that header checks responses.

**Comment by EricSoroos** [comment]
Ok, So I've run a quick check on one of our test systems. 

Null case: No `vary` header, does have `Cache-Control: private` on some pages, public on js assets, and some other no-cache as appropriate.  Login as admin, visit dashboard, logout, back button.  Dashboard is visible.  Similar if we go a dataset/ listing with a private dataset, it's still there. 

Trial: `Vary cookie` in the response. (in this case, I tossed it into nginx because it's faster that way.) Login as admin, visit dashboard, logout, back button. Get a flash error "Not Authorized to see this page" and a redirect to login page. For the dataset page, it reloads and gives me no private datasets. 

![image](https://github.com/ckan/ckan/assets/36156267/31b81e49-2a55-4b48-83bb-70d31ba6f5dc)

**Comment by ThrawnCA** [comment]
Hmm, I must have misread the Vary docs the first time. That does seem promising.

Since the header.html snippet is included everywhere and changes based on login, I'd think that every HTML response could include `Vary: cookie`, leaving only other content types without it.

And perhaps the API should use `Vary: authorization`.

**Comment by EricSoroos** [comment]
Authorization is explicitly part of the cache key according to the docs I read today.

**Comment by wardi** [comment]
@EricSoroos so is authorization in the cache key not working? What do you think of the new fix in #7969 ?

**Comment by EricSoroos** [comment]
from: https://www.rfc-editor.org/rfc/rfc9110#field.vary

> There is no need to send the Authorization field name in Vary because reuse of that response for a different user is prohibited by the field definition ([Section 11.6.2](https://www.rfc-editor.org/rfc/rfc9110#field.authorization)).

I think we might want to vary on 'X-ckan-api', or whatever the alternate authorization header is, since that wouldn't be covered by the above.


And for completeness the same section has this:

> A Vary field containing a list of field names has two purposes:

> 1) To inform cache recipients that they MUST NOT use this response to satisfy a later request unless the later request has the same values for the listed header fields as the original request ([Section 4.1](https://www.rfc-editor.org/rfc/rfc9111#section-4.1) of [[CACHING](https://www.rfc-editor.org/rfc/rfc9110#CACHING)]) or reuse of the response has been validated by the origin server. In other words, Vary expands the cache key required to match a new request to the stored cache entry.

> 2) To inform user agent recipients that this response was subject to content negotiation ([Section 12](https://www.rfc-editor.org/rfc/rfc9110#content.negotiation)) and a different representation might be sent in a subsequent request if other values are provided in the listed header fields ([proactive negotiation](https://www.rfc-editor.org/rfc/rfc9110#proactive.negotiation)).

The solution I'm testing here is exercising part 1 of this -- since the cookie has changed, the browser cache is invalid and it retries.

**Comment by ThrawnCA** [comment]
@EricSoroos We have added `Vary: Cookie` to www.data.qld.gov.au and www.publications.qld.gov.au with great success; customer complaints about seeing stale pages after login have stopped.

**Comment by ThrawnCA** [comment]
Newer AWS application load balancers, when using sticky sessions, will unfortunately update a special cookie on every single request, which makes `Vary: Cookie` infeasible.

**Comment by JVickery-TBS** [comment]
Ian mentioned there was a github issue regarding caching...I think this is it. One of my browsers is fully stuck in this `vary` cookie thing that @duttonw mentioned above. Is there a fix for this? Nothing like clearing browser caches etc works for me sadly.

I have changed the "private" cache to be `no-cache, private` and further made a `nocache_store` blueprint decorator method which sets `no-cache, no-store` on views like the login page, logout, login/logout redirects, register, and password reset.

Though none of this has fixed my login issue with the Vary cookie

**Comment by duttonw** [comment]
Hi @JVickery-TBS ,

See https://github.com/qld-gov-au/ckanext-oidc-pkce/pull/10/files which fixed the oauth login flow from being marked as public cacheable.

See https://github.com/ckan/ckan/pull/8954 which is a major rework of the entire cache platform based on RFC spec's and ETags as well as dropping CSRF tokens from being 'issued' until required meaning pages can be cached far longer.

Big parts are in:
``ckan/views/__init__.py`` function ``def set_cache_control_headers_for_response(response: Response) -> Response:``
``ckan/config/middleware/flask_app.py``
```
   session_access = session.accessed
    # used to mimic session.new since flask_session.new can lie
    g.__session_was_empty = len(session.keys()) == 0
    session.accessed = session_access  # reset session accessed state
```

I've designed this to allow extensions to allow apend extra 'etag' data so we could in the future set the etag to the date modified of the dataset/resource but it still changes the etag if a comment is added, etc.


As to what i've observed in our current modern browsers in relation to AWS Application load balancers:
I really don't know why but AWS ALB with 'sticky session' instead of doing what the legacy system did which was a cookie that rarely changed, they decided to have it 'change' on every _single_ request through the platform. This makes vary by cookie moot as it does seem there is some code deep in the bowels of webkit (chrome/edge) that ignores cache clearing on that vary header for private caching. Current CKAN does not set a max-age on private cache so it keeps it in private cache indefinitely (or until it is manually purged).
